snappy-tools/sbin/collect-debug-artifacts.sh

```diff
@@ -0,0 +1,273 @@
+SCRIPT_DIR="`cd "$SCRIPT_DIR" && pwd`"
+
+while [ "$1" != "" ]; do
+  option="$1"
+
+  case "$option" in
+    -c)
+      CONF_FILE="$2"
+      shift ;;
+    --conf=*|--config=*)
+      CONF_FILE="`echo "$2" | sed 's/^[^=]*=//'`" ;;
+    -h|--help)
+    echo $usage
+    exit 0
+    ;;
+    -a|--all)
+    GET_EVERYTHING=1
+    ;;
+    -v|--verbose)
+    VERBOSE=1
+    ;;
+    -d|--dump)
+    DUMP_STACK=1
+    ;;
+  esac
+  shift # past argument or value
+done
+
+# Check configurations and assign defaults
+function check_configs {
+  if [ -z "${CONF_FILE}" ]; then
+    CONF_FILE="${SCRIPT_DIR}/../conf/debug.conf.template"
+  fi
+
+  if [ ! -f "${CONF_FILE}" ]; then
+    echo "Config file ${CONF_FILE} does not exist"
+    exit 1
+  fi
+
+  source $CONF_FILE
+
+  if [ ! -f "${MEMBERS_FILE}" ]; then
+    echo "members file ${MEMBERS_FILE} does not exist"
+    exit 1
+  fi
+
+  if [ -z "${NO_OF_STACK_DUMPS}" ]; then
+    NO_OF_STACK_DUMPS=5
+  fi
+
+  if [ -z "${INTERVAL_BETWEEN_DUMPS}" ]; then
+    INTERVAL_BETWEEN_DUMPS=5
+  fi
+
+  if [ -z "${GET_EVERYTHING}" ]; then
+    GET_EVERYTHING=0
+  fi
+
+  if [ "${VERBOSE}" = "1" ]; then
+    echo CONF=$CONF_FILE
+    echo MEMINFO=$MEMBERS_FILE
+    echo NUM_STACK_DUMPS=$NO_OF_STACK_DUMPS
+    echo INTERVAL_BETWEEN_DUMPS=$INTERVAL_BETWEEN_DUMPS
+    echo GET_EVERYTHING=${GET_EVERYTHING}
+  fi
+}
+
+collector_host=`hostname`
+
+function collect_data {
+  host="$1"
+  wd="$2"
+  srv_num="$3"
+  top_level_out_dir="$4"
+
+  if [ "${VERBOSE}" = "1" ]; then
+    echo "Collecting data for process running on ${host} with working_dir ${wd}"
+  fi
+
+  # make a sub directory for this host and pid
+  out_dir="${top_level_out_dir}/${host}-${srv_num}"
+
+  mkdir -p $out_dir 
+  if [ "${VERBOSE}" = "1" ]; then
+    echo "Args Being passed"
+    echo "arg1 wd = ${wd}" 
+    echo "arg2 num_stack_dumps = ${NO_OF_STACK_DUMPS}" 
+    echo "arg3 interval_dumps = ${INTERVAL_BETWEEN_DUMPS}" 
+    echo "arg4 GET_EVERYTHING = ${GET_EVERYTHING}" 
+    echo "arg5 collector_host = ${collector_host}" 
+    echo "arg6 out dir = ${out_dir}"
+  fi
+
+  if [ "${DUMP_STACK}" != "1" ]; then
+    NO_OF_STACK_DUMPS=0
+    INTERVAL_BETWEEN_DUMPS=0
+  fi
+
+  typeset -f | ssh $host "$(cat);collect_on_remote \"${wd}\" \"${NO_OF_STACK_DUMPS}\" \\
+      \"${INTERVAL_BETWEEN_DUMPS}\" \"${GET_EVERYTHING}\" \"${collector_host}\" \"${out_dir}\" \"${VERBOSE}\""
+}
+
+function collect_on_remote {
+  data_dir="$1"
+  num_stack_dumps="$2"
+  int_stack_dumps="$3"
+  get_all="$4"
+  collector_host="$5"
+  collector_dir="$6"
+  verbose="$7"
+
+  tmp_dir="$(mktemp -d --tmpdir="`pwd`" data.XXXX)"
+  retval=$?
+  if [ ! -d ${tmp_dir} ]; then
+    echo "FAILED TO CREATE tmp dir on ${host} at ${data_dir} with errno ${retval}"
+  fi
+
+  # first get the pid. The latest log file with the header will have the pid
+  host=`hostname`
+  if [ ! -d "$data_dir" ]; then
+    echo "${data_dir} not found on host: ${host}"
+  fi
+
+  cd "$data_dir"
```

Comment:
since this has cded into data_dir won't usages of tmp_dir will cause trouble below (unless tmp_dir itself is created inside data_dir as mentioned in the comment above (i.e. --tmpdir="$data_dir")